client/client-multi/client-objective-c/Kaa/log/sqlite/SQLiteLogStorage.m

```diff
@@ -0,0 +1,435 @@
+@property (nonatomic) int64_t unmarkedConsumedSize;
+@property (nonatomic) int32_t currentBucketId;
+
+@property (nonatomic) int64_t currentBucketSize;
+@property (nonatomic) int32_t currentRecordCount;
+
+/**
+ * Format: <int32_t, int64_t> as key-value wrapped with NSNumber;
+ * key:     bucket id;
+ * value:   bucket size.
+ */
+@property (nonatomic, strong) NSMutableDictionary *consumedMemoryMap;
+
+- (void)openDBAtPath:(NSString *)path;
+- (void)executeQuery:(NSString *)statement;
+- (sqlite3_stmt *)statementForQuery:(NSString *)query;
+
+- (void)truncateIfBucketSizeIncompatible;
+- (void)retrieveConsumedSizeAndVolume;
+- (void)retrieveBucketId;
+- (void)resetBucketsState;
+- (void)updateStorageParam:(NSString *)param value:(int64_t)value;
+- (void)moveToNextBucket;
+- (void)updateStateForBucketWithId:(int32_t)bucketId;
+
+@end
+
+@implementation SQLiteLogStorage
+
+- (instancetype)initWithBucketSize:(int64_t)bucketSize bucketRecordCount:(int32_t)bucketRecordCount {
+    return [self initWithDatabaseName:DEFAULT_DB_NAME
+                           bucketSize:bucketSize
+                    bucketRecordCount:bucketRecordCount];
+}
+
+- (instancetype)initWithDatabaseName:(NSString *)dbName
+                          bucketSize:(int64_t)bucketSize
+                   bucketRecordCount:(int32_t)bucketRecordCount {
+    self = [super init];
+    if (self) {
+        
+        self.consumedMemoryMap = [NSMutableDictionary dictionary];
+        self.maxBucketSize = bucketSize;
+        self.maxBucketRecordCount = bucketRecordCount;
+        self.currentBucketId = 1;
+        
+        NSString *baseDir = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES)[0];
+        NSString *dbPath = [baseDir stringByAppendingPathComponent:dbName];
+        
+        NSFileManager *fileManager = [NSFileManager defaultManager];
+        if ([fileManager fileExistsAtPath:dbPath]) {
+            DDLogInfo(@"%@ Opening database [%@]", TAG, dbName);
+            [self openDBAtPath:dbPath];
+        } else {
+            NSError *error;
+            BOOL dirCreated = [fileManager createDirectoryAtPath:baseDir withIntermediateDirectories:YES attributes:nil error:&error];
+            if (dirCreated) {
+                DDLogInfo(@"%@ Database [%@] first start!", TAG, dbName);
+                [self openDBAtPath:dbPath];
+            } else {
+                DDLogError(@"%@ Directory [%@] doesn't exist and couldn't be created: %@", TAG, baseDir, error);
+                [NSException raise:KaaRuntimeException format:@"Can't create directory for database"];
+            }
+        }
+    }
+    return self;
+}
+
+-(BucketInfo *)addLogRecord:(LogRecord *)record {
+    @synchronized(self) {
+        DDLogVerbose(@"%@ Adding new log record", TAG);
+        
+        sqlite3_stmt *statement = [self statementForQuery:KAA_INSERT_NEW_RECORD];
+        
+        int64_t leftConsumedSize = self.maxBucketSize - self.currentBucketSize;
```

Comment:
This variable should be named remainingSize or something similar(now it says that size is consumed)